### DISCUSSION TITLE
How mock module with TypeScript?

### DISCUSSION THREAD

**Discussion by cawa-93** [vitest-dev/vitest#444] [Q&A] [answered]
I'm trying to mock the `BrowserWindow` class from the electron module.

This class is used as a constructor `new BrowserWindow(...)` and also has static methods `BrowserWindow.getAllWindows()`.

I wrote mock and tests. And everything works. But I get a warning from the TypeScript, which is logical.
```ts
// TS2339: Property 'mock' does not exist on type 'typeof BrowserWindow'.
expect(BrowserWindow.mock.instances.length).toBe(1);
```

I tryed use `MockedObject` helper. And in part it helped. Static methods are now defined correctly. But there is no access to instance methods:

```ts
// @ts-expect-error
const BrowserWindowMocked = BrowserWindow as MockedObject<typeof BrowserWindow>;

expect(BrowserWindowMocked.getAllWindows.mock.calls.length).toEqual(0); // Ok

//  Property 'loadURL' does not exist on type '[options?: BrowserWindowConstructorOptions]'.
expect(BrowserWindowMocked.mock.instances[0].loadURL).toHaveBeenCalledTimes(1);
```

My full demo can be founded here: https://github.com/cawa-93/vite-electron-builder/blob/3d81e3793da77ba56f40f5a5ad3531feb9568ee8/packages/main/tests/unit.spec.ts

**Comment by cawa-93** [comment, marked as answer]
Type below works for me
```ts
type TBrowserWindowMocked = typeof BrowserWindow
  & MockedObjectDeep<typeof BrowserWindow>
  & JestMockCompatFn<ConstructorParameters<typeof BrowserWindow>, MockedObjectDeep<BrowserWindow>>;

const BrowserWindowMocked = BrowserWindow as TBrowserWindowMocked 
```
Full test may be found here: https://github.com/cawa-93/vite-electron-builder/blob/cafa809218ae52a311beaee3bcc86f97abb3fabe/packages/main/tests/unit.spec.ts

**Reply by sheremet-va** [reply]
We have a type hint method for that:

You can use `vi.mocked(BrowserWindow).mock.instances.length`

**Reply by cawa-93** [reply]
Awesome. Thanks

**Reply by cawa-93** [reply]
It seems that this helper does not mock the properties of the class instance:
```ts
// Property 'mockReturnValueOnce' does not exist on type '(url: string, options?: LoadURLOptions) => Promise '.
vi.mocked(BrowserWindow, true).mock.instances[0].loadURL.mockReturnValueOnce(true)
```

I fix this by adding:
```ts
vi.mocked(BrowserWindow as typeof BrowserWindow
  & JestMockCompatFn<ConstructorParameters<typeof BrowserWindow>, MaybeMockedDeep<BrowserWindow>> )

**Reply by cawa-93** [reply]
... or just use `vi.mocked` twice:
```ts
const mockedBW = vi.mocked(BrowserWindow)
const instance = vi.mocked(mockedBW.mock.instances[0])
```
